_posts/2022-02-17-passwordless-auth-with-express-and-postgres.md

```diff
@@ -0,0 +1,563 @@
+
+So why one-time passwords? First of all they're good for users because they don't have to remember a password. Just put your email address into the login field and they'll be sent an auto-generated password or link to log them in. Secondly these passwords are short-lived, which means that if a data breach _does_ happen then there are no active passwords stored for crackers to log in with or share, which reduces the attack window dramatically. As long as a user has access to their email address their accounts will be safe.
+
+# Disclaimer
+
+This article assumes you know your way around PostgreSQL and ExpressJS. We will not be hand-holding you through the entire application build, but instead will be focusing on the concepts and things you'll need to know to implement a safe and secure one-time password login system.
+
+A complete working example will be available at the end of the tutorial.
+
+# Overview
+
+For this article we'll be implementing the most simple version of a one-time password (OTP): a short code sent via email. This was chosen for simplicity, availability, and cost. Almost every user will have an email, and there is very minimal cost in sending them compared with something like SMS. A OTP is used here over a link also for user convenience. It allows them to retrieve the code on one device (like a phone) and log into your service on another (like a public computer).
+
+As a high-level overview the implementation will be as follows:
+
+- A user registers an account using their email address
+- The user attempts to log in using their email address
+- The site retrieves the user information and generates a OTP
+- The OTP is stored securely in the database, and sent via plain text to the user
+- The user will then retrieve the code and enter it into the login field provided
+- The OTP is verified against the database
+- If successful, the user is logged into the application 
+- The OTP is then deleted from the database to prevent re-use
+
+This flow is flexible and can be implemented using any technology you are familiar with, but as mentioned we'll be using PostgreSQL and ExpressJS for this example.
+
+# Implementation
+
+## Setup
+
+Generate yourself an Express app, and set up your Postgres database. For libraries we'll be using:
+
+- [node-postgres](https://github.com/brianc/node-postgres) for interfacing with our database
+- [node-jsonwebtoken](https://github.com/auth0/node-jsonwebtoken) for the JWT
+- [bcrypt](https://github.com/kelektiv/node.bcrypt.js) for encrypting the one-time passwords
+- [emailjs](https://github.com/eleith/emailjs) for emailing the user
+- [dotenv](https://github.com/motdotla/dotenv) for loading our secrets and parameters
+
+For our `.env` file we need the following environment variables in place:
+
+```env
+# Database connection string
+DATABASE_URL=postgresql://username:password@localhost:5432/dbname
+# Secret key for secure JWT generation
+JWT_SECRET_KEY=ARandomStringOfCharacters0389j0jsdf8jr8h9as8jd
```

Comment:
It might be worth pointing out that for a live system you would ideally put secrets in an encrypted store of some kind.

Reply:
Done